### PR TITLE
add setup hook capabilities for rke2

### PR DESCRIPTION
**Pull request by briandowns**
Proposed changes
======
This PR adds the ability to add hooks to be run in bootstrapping in k3s that will ultimately be used during rke2 startup.

Types of changes
======
Adds a slice to a struct field and referenced during k3s start up.

Verification
======
Verification will be had through it's use in this RKE2 pull request. https://github.com/rancher/rke2/pull/199

Linked Issues
======
N/A

Further comments
======
N/A

### CODE REVIEW

**Comment by briandowns** [comment]
They'd be used during RKE2 setup. I open to calling them anything that makes their use clear.

**Comment by brandond** [comment]
So k3s calls these hooks during startup, but rke2 calls them during setup? Or does rke2 add them during setup? I'm in favor of naming them after the place they get called.

**Comment by briandowns** [comment]
changed

File: pkg/server/server.go [hunk 1]
```diff
@@ -60,6 +60,10 @@ func StartServer(ctx context.Context, config *Config) error {
 		return errors.Wrap(err, "starting tls server")
 	}
 
+	for _, hook := range config.StartupHooks {
+		hook(config.ControlConfig)
```
Comment (cjellick) [inline pkg/server/server.go:64]: I think the hook function should return an error and it should be wrapped and return if not nil

Comment (cjellick) [reply pkg/server/server.go:64]: oh, the function already does but this doesnt check them

**Comment by briandowns** [comment]
Updated. Thanks @cjellick @dweomer @brandond

**Pull request merged**